File: packages/nodejs/src/auth/oauth.ts
import type { Metadata } from '@grpc/grpc-js';
import axios from 'axios';
import type { Cache, Store, MemoryStore } from 'cache-manager';
import { memoryStore } from 'cache-manager';
import type { VMXClient } from '../client';
import type { VMXClientAuthProvider } from './types';

export type VMXClientOAuthOptions<TCacheStore extends Store = MemoryStore> = {
  clientId: string;
  clientSecret: string;
  cacheManager?: Cache<TCacheStore>;
};

type OAuthTokenResult = {
  access_token: string;
  expires_in: number;
  token_type: string;
};

export class VMXClientOAuth<TCacheStore extends Store = MemoryStore> implements VMXClientAuthProvider {
  private cache: Cache<TCacheStore> | MemoryStore;

  constructor(private readonly options: VMXClientOAuthOptions<TCacheStore>) {
    this.cache = options.cacheManager || memoryStore();
  }

  public async injectCredentials(client: VMXClient, grpcMetadata: Metadata): Promise<void> {
    const token = await this.getOAuthToken(client.options.domain);

    grpcMetadata.add('Authorization', `Bearer ${token}`);
  }

  private async getOAuthToken(domain: string): Promise<string> {
    const token = await this.cache.get<string>('oauth_token');
    if (token) {
      return token;
    }

    const response = await axios.post(
      `https://auth.${domain}/oauth2/token`,
      {
        grant_type: 'client_credentials',
        client_id: this.options.clientId,
        client_secret: this.options.clientSecret,
      },
      {
        headers: {
          'Content-Type': 'application/x-www-form-urlencoded',
        },
      },
    );

    const result = response.data as OAuthTokenResult;

    this.cache.set('oauth_token', result.access_token, (result.expires_in - 5) * 1000);

    return result.access_token;
  }
}

File: packages/python/vmxai/auth/oauth.py
from typing import TYPE_CHECKING, Generic, Optional, TypedDict, TypeVar

from cachetools import Cache, TTLCache
from requests import post

from vmxai.auth.provider import VMXClientAuthProvider

if TYPE_CHECKING:
    from vmxai.client import VMXClient

T = TypeVar("T", bound=Cache)


class OAuthTokenResult(TypedDict):
    access_token: str
    expires_in: int
    token_type: str


class VMXClientOAuth(Generic[T], VMXClientAuthProvider):
    def __init__(self, client_id: str, client_secret: str, cache_manager: Optional[T] = None):
        self.client_id = client_id
        self.client_secret = client_secret
        self.cache = cache_manager or TTLCache(
            maxsize=100,
            ttl=3540,
        )  # 59 minutes

    def inject_credentials(self, client: "VMXClient", grpc_metadata: list[tuple[str, str]]) -> tuple[tuple[str, str]]:
        token = self.get_oauth_token(client.domain)
        grpc_metadata.append(("authorization", f"Bearer {token}"))
        return grpc_metadata

    def get_oauth_token(self, domain: str) -> str:
        token = self.cache.get("oauth_token")
        if token:
            return token

        response = post(
            f"https://auth.{domain}/oauth2/token",
            data={
                "grant_type": "client_credentials",
                "client_id": self.client_id,
                "client_secret": self.client_secret,
            },
            headers={
                "Content-Type": "application/x-www-form-urlencoded",
            },
        )

        result: OAuthTokenResult = response.json()
        self.cache["oauth_token"] = result["access_token"]

        return result["access_token"]
